enhancements/ingress/forwarded-header-policy.md

```diff
@@ -0,0 +1,339 @@
+
+#### As a cluster administrator, I have configured an external proxy that injects the `X-Forwarded-For` header into each request before forwarding it to an IngressController, and I want the IngressController to pass the header through unmodified.
+
+To satisfy this use-case, the cluster administrator can specify the "never"
+policy.  The IngressController then never sets the headers, and applications
+receive only the headers that the external proxy provides.
+
+#### As a cluster administrator, I want an IngressController to pass the `X-Forwarded-For` header that my external proxy sets on extra-cluster requests through unmodified, and I want the IngressController to set the `X-Forwarded-For` header on intra-cluster requests, which do not go through the external proxy.
+
+To satisfy this use-case, the cluster administrator can specify the "if-none"
+policy.  If an HTTP request already has the header (presumably set the external
+proxy), the IngressController preserves it.  If the header is absent because the
+request did not come through the proxy, then the IngressController adds the
+header.
+
+#### As an application developer, I have configured an application-specific external proxy that injects the `X-Forwarded-For` header, and I want an IngressController to pass the header through unmodified for my application's Route, without affecting the policy for other Routes.
+
+To satisfy this use-case, the application developer can add an annotation
+`haproxy.router.openshift.io/set-forwarded-headers: if-none` or
+`haproxy.router.openshift.io/set-forwarded-headers: never` on the Route for the
+application.
+
+### Implementation Details
+
+Implementing this enhancement requires changes in the following repositories:
+
+* openshift/api
+* openshift/cluster-ingress-operator
+* openshift/router
+
+The router configures HAProxy using a configuration template.  The template uses
+environment variables as input parameters.  The enhancement adds a new
+environment variable, `ROUTER_SET_FORWARDED_HEADERS`, which specifies the
+policy.  The ingress operator sets this variable on the router's deployment
+based on the IngressController's `spec.httpHeaders.forwarded.policy` field
+value.
+
+HAProxy has an `option forwardfor` configuration keyword to specify when and how
+HAProxy sets the `X-Forwarded-For` header, as well as `http-request add-header`
+and `http-request set-header` keywords to append or replace arbitrary headers.
+The `option forwardfor` keyword has an associated `if-none` keyword, which can
+be used to implement the homonymous IngressController policy.  The `http-request
+set-header` keyword can be conditionalized using ACLs to implement `if-none`.
+These keywords are sufficient to implement all four policies.
+
+The HAProxy configuration template generates a backend configuration block for
+each route based on the value in the route's annotation or the value of
+`ROUTER_SET_FORWARDED_HEADERS`.  To implement `append` for the `X-Forwarded-For`
```

Comment:
> based on the value in the route's annotation or the value of
`ROUTER_SET_FORWARDED_HEADERS`.

If the route's annotation and `ROUTER_SET_FORWARDED_HEADERS` specify a forwarded header policy, the route's annotation will override `ROUTER_SET_FORWARDED_HEADERS`, correct?

Reply:
Correct.